vendor/github.com/ethereum/go-ethereum/whisper/whisperv5/whisper.go

```diff
@@ -718,6 +735,33 @@ func (wh *Whisper) add(envelope *Envelope) (bool, error) {
 	return true, nil
 }
 
+func (w *Whisper) traceIncomingDelivery(isP2P bool, status message.Status, src NewMessage, env *Envelope, rec *ReceivedMessage, err error) {
+	w.traceDelivery(isP2P, message.IncomingMessage, status, src, env, rec, err)
+}
+
+func (w *Whisper) traceOutgoingDelivery(isP2P bool, status message.Status, src NewMessage, env *Envelope, rec *ReceivedMessage, err error) {
+	w.traceDelivery(isP2P, message.OutgoingMessage, status, src, env, rec, err)
+}
+
+func (w *Whisper) traceDelivery(isP2P bool, dir message.Direction, status message.Status, src NewMessage, env *Envelope, rec *ReceivedMessage, err error) {
+	if w.deliveryServer != nil {
+		var envelope Envelope
+
+		if env != nil {
+			envelope = *env
+		}
```

Comment:
Probably better to add dereferencing also for `*ReceivedMesage` parameter, and change `NewMessage` to pointer as well, so it's consistant with other struct paramters - easier for callers.